src/scheduler.rs

```diff
@@ -0,0 +1,395 @@
+                    base_psbt: Vec::new(),
+                    no_publish: true,
+                };
+                let funding_shim = tonic_lnd::rpc::funding_shim::Shim::PsbtShim(psbt_shim);
+                let funding_shim = tonic_lnd::rpc::FundingShim { shim: Some(funding_shim) };
+
+                let req = OpenChannelRequest {
+                    node_pubkey: chan.node.node_id.to_vec(),
+                    local_funding_amount: chan
+                        .amount
+                        .as_sat()
+                        .try_into()
+                        .expect("amount too large"),
+                    push_sat: 0,
+                    private: false,
+                    min_htlc_msat: 0,
+                    remote_csv_delay: 0,
+                    spend_unconfirmed: false,
+                    close_address: String::new(),
+                    funding_shim: Some(funding_shim),
+                    remote_max_value_in_flight_msat: chan.amount.as_sat() * 1000,
+                    remote_max_htlcs: 10,
+                    max_local_csv: 288,
+                    ..Default::default()
+                };
+
+                (node_addr, chan_id, req)
+            })
+            .collect()
+    }
+
+    // gen_funding_created AKA
+    fn add_channels_to_psbt<I>(
+        &self,
+        original_psbt: PartiallySignedTransaction,
+        owned_vout: usize,
+        funding_txos: I,
+    ) -> PartiallySignedTransaction
+    where
+        I: IntoIterator<Item = TxOut>,
+    {
+        let mut iter = funding_txos.into_iter();
+        let funding_txout = iter.next().unwrap(); // we assume there is at least 1.
+
+        let mut proposal_psbt = original_psbt.clone();
+        // determine whether we replace original psbt's owned output
+        // or whether we change the value to be wallet amount
+        if self.wallet_amount() == bitcoin::Amount::ZERO {
+            assert_eq!(funding_txout.value, self.channels[0].amount.as_sat());
+            proposal_psbt.unsigned_tx.output[owned_vout] = funding_txout;
+        } else {
+            proposal_psbt.unsigned_tx.output[owned_vout].value = self.wallet_amount().as_sat();
+            proposal_psbt.unsigned_tx.output.push(funding_txout)
+        }
+
+        // add remaining funding txouts
+        proposal_psbt.unsigned_tx.output.extend(iter);
+
+        // psbt should have outputs same length as contained unsigned tx
+        proposal_psbt.outputs.resize_with(proposal_psbt.unsigned_tx.output.len(), Default::default);
+
+        eprintln!("channel funding Proposal PSBT created: {:#?}", proposal_psbt);
+
+        proposal_psbt
+    }
+}
+
+pub type ChannelId = [u8; 32];
+
+fn temporary_channel_id() -> ChannelId {
+    rand::random()
+}
+
+/// [Scheduler] lets you pre-batch PayJoins, usually containing Channel Opens.
+#[derive(Clone)]
+pub struct Scheduler {
+    lnd: LndClient,
+    pjs: Arc<Mutex<HashMap<Script, ScheduledPayJoin>>>, // payjoins mapped by owned `scriptPubKey`s
```

Comment:
I don't like the idea of `ScheduledPayJoin`. We should just have a list of channels we want to open (funding amount + node we want to open with). And when payjoin requests come, we check the amount, and determine if it is enough to open a channel (or multiple), by looking at our list of "scheduled channels".

Maybe this is out of scope of this PR (since this is just a refactor so the codebase becomes more maintainable), but I think this would be a decent change.

Reply:
How do you determine the amount to put in the bip21? Just allow it to be arbitrary?

Reply:
So we would have 2 admin endpoints: add scheduled channel, add bip-21 request.

"Add bip-21 request" will add to a map that associates recipient (owned) `scriptPubKey` to "payment constraints" (i.e. minimum acceptable amount, maximum acceptable amount, etc.) so when we receive a payment, we know whether we have requested it, and whether the payment satisfies requested amount.

Now when we actually receive a payment request (via an incoming bip-078 original psbt), we check:
1. Whether the psbt contains outputs that match our recipient `scriptPubKey`s. If one or more do, we need to ensure that the outputs that match satisfy the value constraints. If not, we need to reject the payment (I'm not sure how this is usually done). If we deem the outputs okay, continue to 2.
2. We need to check if the original psbt actually creates a valid tx (because this is our fallback in case we fail payjoin).
3. We sum up the values of "owned" outputs of the psbt. This is the max value we can work with when opening channels. We check how many scheduled channels we can open given this value.
4. We initiate opens of these picked scheduled channels and create proposal psbt that we send back to sender.